### PR TITLE
Fix Issure 2286 - RyuJIT needs to honor volatile reads/writes and avoid making CSE's

### PR DESCRIPTION
**Pull request by briansull**
RyuJIT needs to honor volatile reads/writes and avoid making CSE's across them

In Liveness.cpp we will now treat a volatile accesses as definitions of the global heap
Assertions are also added in ValueNumber.cpp such that a block that defines or mutates the global heap
must have been marked as containing a definition of the global heap using the bbHeapDef flag.

### CODE REVIEW

**Comment by briansull** [comment]
@dotnet/jit-contrib Fixes for issue #2286 .

File: src/jit/valuenum.cpp [hunk 1]
```diff
@@ -5516,7 +5525,12 @@ void Compiler::fgValueNumberTree(GenTreePtr tree, bool evalAsgLhsInd)
                         }
                         else
                         {
-                            fgMutateHeap(tree DEBUGARG("assign-of-IND"));
+                            GenTreeLclVarCommon* dummyLclVarTree = NULL;
+                            if (!tree->DefinesLocal(this, &dummyLclVarTree))
+                            {
+                                // If it doesn't define a local, then it might update the heap.
+                                fgMutateHeap(tree DEBUGARG("assign-of-IND"));
```
Comment (CarolEidt) [inline src/jit/valuenum.cpp]: For this case, should there be an assert(compCurBB->bbHeapDef)?

**Comment by CarolEidt** [comment]
Other than my one question about a possible additional assert, LGTM.

File: src/jit/valuenum.cpp (hunk 1 above)
Comment (briansull) [inline src/jit/valuenum.cpp]: See line 4433. there is such an assert in the body of fgMutateHeap()

Comment (CarolEidt) [inline src/jit/valuenum.cpp]: Great thanks!

File: src/jit/valuenum.cpp [hunk 2]
```diff
@@ -5516,7 +5525,12 @@ void Compiler::fgValueNumberTree(GenTreePtr tree, bool evalAsgLhsInd)
                         }
                         else
                         {
-                            fgMutateHeap(tree DEBUGARG("assign-of-IND"));
+                            GenTreeLclVarCommon* dummyLclVarTree = NULL;
```
Comment (sivarv) [inline src/jit/valuenum.cpp]: CC: nullptr

**Comment by sivarv** [comment]
LGTM, other than the coding convention nit.

**Pull request merged**